File: [9.1] Хэш-таблица/[9.1] Хэш-таблица/list.cpp
#include <iostream>
#include <string>
#include "list.h"

struct ListElement
{
	std::string word;
	int value;
	ListElement *next;
};

ListElement *createListElement(const std::string &word, int value, ListElement *next)
{
	ListElement *newElement = new ListElement;
	newElement->next = next;
	newElement->value = value;
	newElement->word = word;
	return newElement;
}

bool isEmpty(ListElement *head)
{
	return head == nullptr;
}

int amountOfElements(ListElement *head)
{
	int count = 0;
	ListElement *cursor = head;
	while (cursor != nullptr)
	{
		++count;
		cursor = cursor->next;
	}
	return count;
}

bool isContained(ListElement *head, const std::string &word)
{
	ListElement *cursor = head;
	while (cursor != nullptr)
	{
		if (cursor->word == word)
		{
			return true;
		}
		cursor = cursor->next;
	}
	return false;
}

void addElement(ListElement *head, const std::string &word)
{
	head = createListElement(word, 1, head);
}

void increaseNumber(ListElement *head, const std::string &word)
{
	ListElement *cursor = head;
	while (cursor->word != word)
	{
		cursor = cursor->next;
	}
	++cursor->value;
}

void printList(ListElement *head)
{
	ListElement *cursor = head;
	while (cursor != nullptr)
	{
		std::cout << " ' " << cursor->word << " ' (" << cursor->value << ") ";
		cursor = cursor->next;
	}
}

void deleteList(ListElement *head)
{
	ListElement *toDelete = head;
	while (head != nullptr)
	{
		head = head->next;
		delete toDelete;
		toDelete = nullptr;
	}
}

File: semester_2/hw2/Task_3/Task_3/List.cs
using System;

namespace Task_3
{
    /// <summary>
    /// Класс связный список
    /// </summary>
    public class List
    {
        /// <summary>
        /// Класс элемент связного списка
        /// </summary>
        private class ListElement
        {
            /// <summary>
            /// Указатель на следующий элемент
            /// </summary>
            public ListElement Next { get; set; }

            /// <summary>
            /// Значение элемента
            /// </summary>
            public string Value { get; private set; }

            /// <summary>
            /// Конструктор для класса ListElement
            /// </summary>
            /// <param name="next">Указатель на следующий элемент</param>
            /// <param name="value">Значение элемента</param>
            public ListElement(ListElement next, string value)
            {
                this.Next = next;
                this.Value = value;
            }

        }

        /// <summary>
        /// Указатель на первый элемент списка
        /// </summary>
        private ListElement head;

        /// <summary>
        /// Длина списка
        /// </summary>
        private int length;

        /// <summary>
        /// Добавление нового элемента в начало списка
        /// </summary>
        /// <param name="value">Значение для добавления</param>
        public void Add(string value)
        {
            var newElement = new ListElement(head, value);
            head = newElement;
            ++length;
        }

        /// <summary>
        /// Проверка списка на пустоту
        /// </summary>
        /// <returns>True, если список пуст, иначе false</returns>
        public bool IsEmpty() => length == 0;

        /// <summary>
        /// Получение длины списка
        /// </summary>
        /// <returns>Длина списка</returns>
        public int GetLength() => length;

        /// <summary>
        /// Проверка значения на принадлежность списку
        /// </summary>
        /// <param name="value">Значение для поиска</param>
        /// <returns>True, если значение принадлежит списку, иначе false</returns>
        public bool IsContained(string value)
        {
            var cursor = head;

            for (var i = 0; i < length; ++i)
            {
                if (cursor.Value == value)
                {
                    return true;
                }
                cursor = cursor.Next;
            }

            return false;
        }

        /// <summary>
        /// Удаление элемента из списка
        /// </summary>
        /// <param name="value">Значение для удаления</param>
        public void RemoveElement(string value)
        {
            if (IsEmpty())
            {
                throw new EmptyListException("List is empty!");
            }

            if (!IsContained(value))
            {
                throw new NotContainedValueException("The word is not contained in the list!");
            }

            if (head.Value == value)
            {
                head = head.Next;
                --length;
                return;
            }

            var cursor = head;
            for (var i = 0; i < length - 1; ++i)
            {
                if (cursor.Next.Value == value)
                {
                    cursor.Next = cursor.Next.Next;
                    --length;
                    return;
                }

                cursor = cursor.Next;
            }
        }

        /// <summary>
        /// Получение первого элемента списка
        /// </summary>
        /// <returns>Значение первого элемента списка</returns>
        public string Peek()
        {
            if (IsEmpty())
            {
                throw new EmptyListException("List is empty");
            }

            return head.Value;
        }
    }
}
